### ISSUE TITLE
Customising Jupyter with custom.js, etc.

### ISSUE THREAD

**Issue by a1ex5** [rationalmatter/Juno-Issue-Tracker#29]
Since Jupyter’s config files are buried in an ApplicationSupport folder somewhere, could we have a user-accessible location for `custom.js` and so on? Like the `site-packages` folder. 

Really enjoying the beta, though. Great stuff.

**Comment by navoshta** [comment]
I will consider allowing _some_ configuration, although generally I wouldn't expect direct access to any of the Jupyter internals or config files. 

Generally, my idea is to make the experience as user-friendly as possible, which also means I don't want to let the user poke around the internals _too much_, potentially breaking app's core functionality.

**Comment by a1ex5** [comment]
Absolutely. My idea was to simply to have a “user” folder, like 'site-packages', where users could put stuff. Currently, for example, a 'custom.js' file would need to be installed at 

'/private/.../Application Support/com.rationalmatter.junoapp/python-home/.jupyter/custom/custom.js'

which nobody wants, especially because that’s the _system_ directory. If a 'jupyter/...' directory (without the dot) in the user’s Juno folder presents too much scope for idiocy, you could keep it the dot, and add a little import UI to put files in the right places on behalf of the user. AFAIK, “configuring the server” is just a matter of pointing it at the right folder. 

I’m not suggesting that the app would be useless without this, just that it ought to be fairly simple to enable a bit of customisation, the odd notebook (_not_ kernel) extension, and so on.

À propos package requests, can I just tell you how much I’m enjoying mpmath as a SciPy replacement. In areas like radiometry, the reduced performance is more than made up for by not having to fuss over floating point precision every five minutes!

**Comment by capaldo** [comment]
Any update on this? I really enjoy using the app but was hoping for the ability to customize/have more granular control.

**Comment by navoshta** [comment]
Sorry @capaldo — no news on this one yet. I wouldn't rule out allowing notebook UI customisation in some obscure way in future, but I still don't have a good plan on how this will work. It may be a bit more complicated than it seems, as these customisations would undoubtedly clash with some of Juno's own ones under the hood, and Juno won't be very helpful in telling you what exactly got broken.
